demoapp/src/main/java/com/transferwise/tasks/demoapp/examples/RateLimitingTaskConcurrencyPolicy.java

```diff
@@ -0,0 +1,41 @@
+package com.transferwise.tasks.demoapp.examples;
+
+import com.google.common.util.concurrent.RateLimiter;
+import com.transferwise.tasks.domain.IBaseTask;
+import com.transferwise.tasks.handler.SimpleTaskConcurrencyPolicy;
+import com.transferwise.tasks.handler.interfaces.ITaskConcurrencyPolicy;
+import java.time.Instant;
+import lombok.NonNull;
+
+public class RateLimitingTaskConcurrencyPolicy implements ITaskConcurrencyPolicy {
+
+  // Up to 5 tasks in every second
+  private RateLimiter rateLimiter = RateLimiter.create(5);
+
+  // But not more than 3 in parallel
+  private ITaskConcurrencyPolicy concurrencyLimitingPolicy = new SimpleTaskConcurrencyPolicy(3);
+
+  @Override
+  public @NonNull BookSpaceResponse bookSpace(IBaseTask task) {
+    BookSpaceResponse concurrencyLimitingResponse = concurrencyLimitingPolicy.bookSpace(task);
+    if (!concurrencyLimitingResponse.isHasRoom()) {
+      return concurrencyLimitingResponse;
+    }
+
+    if (rateLimiter.tryAcquire()) {
+      return concurrencyLimitingResponse;
+    }
+    concurrencyLimitingPolicy.freeSpace(task);
+
+    // Most optimal would be to set `tryAgainTime` when a new permit becomes available.
```

Comment:
indeed, that's a shame. all we basically need is to do a reservation in the future and get when this future is.
which is exactly what `reserveAndGetWaitLength` does